Runtime and endpoint reference (#386)
opendatahub-io/opendatahub-documentation: modules/ref-inference-endpoints.adoc

--

An example manifest is available in the link:https://github.com/opendatahub-io/caikit-tgis-serving/blob/main/demo/kserve/custom-manifests/caikit/caikit-standalone/caikit-standalone-servingruntime-grpc.yaml[caikit-tgis-serving GitHub repository^].
endif::[]

TGIS Standalone ServingRuntime for KServe::

* `:443 fmaas.GenerationService/Generate`
* `:443 fmaas.GenerationService/GenerateStream`
+
[NOTE]
--
To query the endpoint for the TGIS standalone runtime, you must also download the files in the link:https://github.com/opendatahub-io/text-generation-inference/blob/main/proto[proto^] directory of the {productname-short} `text-generation-inference` repository.
--

OpenVINO Model Server::

* `/v2/models/<model-name>/infer`

vLLM ServingRuntime for KServe::

* `:443/version`
* `:443/docs`
* `:443/v1/models`
* `:443/v1/chat/completions`
* `:443/v1/completions`
* `:443/v1/embeddings`

> +

Reviewer: We need to add tokenize and detokenize endpoint as well . reference : https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/api_server.py#L165-L176 

` /tokenize, Methods: POST
: /detokenize, Methods: POST```

Author: Do you all agree with this? @dtrifiro @terrytangyuan @christinaexyou

Reviewer 2: Yes, tokenize/detokenize endpoints were added in [v0.5.1](https://github.com/vllm-project/vllm/releases/tag/v0.5.1),  `#5054`, `https://github.com/vllm-project/vllm/pull/5054`

> * `:443/v1/embeddings`
> +

Author: ```suggestion
* `:443/v1/embeddings`
* `:443/tokenize`
* `:443/detokenize`
+
```
@dtrifiro @tarukumar Let me know if this looks right

Reviewer: looks good

:_module-type: REFERENCE

[id='inference-endpoints_{context}']
= Inference endpoints

[role='_abstract']
These examples show how to use inference endpoints to query the model.

Caikit TGIS ServingRuntime for KServe::

* `:443/api/v1/task/text-generation`
* `:443/api/v1/task/server-streaming-text-generation`

> // * `:443/api/v1/task/text-classification`
> // * `:443/api/v1/task/token-classification`

Reviewer 2: ```suggestion
```

Caikit Standalone ServingRuntime for KServe::

If you are serving multiple models, you can query `/info/models` or `:443 caikit.runtime.info.InfoService/GetModelsInfo` to view a list of served models.
+
--
.REST endpoints

* `/api/v1/task/embedding`
* `/api/v1/task/embedding-tasks`
* `/api/v1/task/sentence-similarity`
* `/api/v1/task/sentence-similarity-tasks`
* `/api/v1/task/rerank`
* `/api/v1/task/rerank-tasks`
* `/info/models`
* `/info/version`
* `/info/runtime`
--
+
--
.gRPC endpoints

* `:443 caikit.runtime.Nlp.NlpService/EmbeddingTaskPredict`
* `:443 caikit.runtime.Nlp.NlpService/EmbeddingTasksPredict`
* `:443 caikit.runtime.Nlp.NlpService/SentenceSimilarityTaskPredict`
* `:443 caikit.runtime.Nlp.NlpService/SentenceSimilarityTasksPredict`
* `:443 caikit.runtime.Nlp.NlpService/RerankTaskPredict`
* `:443 caikit.runtime.Nlp.NlpService/RerankTasksPredict`
* `:443 caikit.runtime.info.InfoService/GetModelsInfo`
* `:443 caikit.runtime.info.InfoService/GetRuntimeInfo`
--

ifdef::upstream[]
[NOTE]
--

> By default, the Caikit Standalone Runtime exposes REST endpoints for use. To use gRPC protocol, manually deploy a custom Caikit Standalone ServingRuntime. For more information, see link:{odhdocshome}/serving-models/#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform]. 

Reviewer 3: You could perhaps delete "for use" at the end of the first sentence.

Author: ```suggestion
By default, the Caikit Standalone Runtime exposes REST endpoints. To use gRPC protocol, manually deploy a custom Caikit Standalone ServingRuntime. For more information, see link:{odhdocshome}/serving-models/#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform]. 
```

> By default, the Caikit Standalone Runtime exposes REST endpoints for use. To use gRPC protocol, manually deploy a custom Caikit Standalone ServingRuntime. For more information, see link:{odhdocshome}/serving-models/#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform]. 

Reviewer 3: Line 51, second sentence. Are we missing the word "the" before gRPC protocol, or is it just the way I'm reading it?

Author: I'm treating this the same way as "REST" since they're both protocol. If we should add "the" before gRPC, we should also add "the" before "REST." Does that make sense? Does that change anything for you here?

--

An example manifest is available in the link:https://github.com/opendatahub-io/caikit-tgis-serving/blob/main/demo/kserve/custom-manifests/caikit/caikit-standalone/caikit-standalone-servingruntime-grpc.yaml[caikit-tgis-serving GitHub repository^].
endif::[]

ifndef::upstream[]
[NOTE]
--

> By default, the Caikit Standalone Runtime exposes REST endpoints for use. To use gRPC protocol, manually deploy a custom Caikit Standalone ServingRuntime. For more information, see link:{rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform].

Reviewer 3: See previous comment about "for use".

Reviewer 3: and the previous comment about the gPRC protocol.

Author: ```suggestion
By default, the Caikit Standalone Runtime exposes REST endpoints. To use gRPC protocol, manually deploy a custom Caikit Standalone ServingRuntime. For more information, see link:{rhoaidocshome}{default-format-url}/serving_models/serving-large-models_serving-large-models#adding-a-custom-model-serving-runtime-for-the-single-model-serving-platform_serving-large-models[Adding a custom model-serving runtime for the single-model serving platform].
```

--

An example manifest is available in the link:https://github.com/opendatahub-io/caikit-tgis-serving/blob/main/demo/kserve/custom-manifests/caikit/caikit-standalone/caikit-standalone-servingruntime-grpc.yaml[caikit-tgis-serving GitHub repository^].
endif::[]

TGIS Standalone ServingRuntime for KServe::

* `:443 fmaas.GenerationService/Generate`
* `:443 fmaas.GenerationService/GenerateStream`
+
[NOTE]
--

> To query the endpoint for the TGIS standalone runtime, you must also download the files in the link:https://github.com/opendatahub-io/text-generation-inference/blob/main/proto[proto^] directory of the Open Data Hub `text-generation-inference` repository.

Reviewer 3: Instead of hard-coding "Open Data Hub", use {productname-short} instead.

Author: ```suggestion
To query the endpoint for the TGIS standalone runtime, you must also download the files in the link:https://github.com/opendatahub-io/text-generation-inference/blob/main/proto[proto^] directory of the {productname-short} `text-generation-inference` repository.
```

----

*TGIS Standalone ServingRuntime for KServe*
[source]
----
grpcurl -proto text-generation-inference/proto/generation.proto -d '{"requests": [{"text":"<text>"}]}' -H 'Authorization: Bearer <token>' -insecure <inference_endpoint_url>:443 fmaas.GenerationService/Generate 
----

*OpenVINO Model Server*
[source]
----
curl -ks <inference_endpoint_url>/v2/models/<model_name>/infer -d '{ "model_name": "<model_name>", "inputs": [{ "name": "<name_of_model_input>", "shape": [<shape>], "datatype": "<data_type>", "data": [<data>] }]}' -H 'Authorization: Bearer <token>'
----

*vLLM ServingRuntime for KServe*
[source]
----
curl -v https://<inference_endpoint_url>:443/v1/chat/completions -H "Content-Type: application/json" -d '{ "messages": [{ "role": "<role>", "content": "<content>" }] -H 'Authorization: Bearer <token>'
----
--
endif::[]

[role='_additional-resources']
== Additional resources
* link:https://github.com/IBM/text-generation-inference[Text Generation Inference Server (TGIS)^]
* link:https://caikit.readthedocs.io/en/latest/autoapi/caikit/index.html[Caikit API documentation^]

> * link:https://github.com/markstur/caikit-embeddings[Caikit Text Embedding GitHub project^]

Reviewer 2: ```suggestion
* link:https://github.com/caikit/caikit-nlp[Caikit NLP GitHub project^]
```

* link:https://docs.openvino.ai/2023.3/ovms_docs_rest_api_kfs.html[OpenVINO KServe-compatible REST API documentation^]

> * link:https://platform.openai.com/docs/api-reference/introduction[OpenAI API documentation]^

Reviewer 2: Not sure if we want to directly link to openai. Maybe it's worth it to document that the api is self-documenting (using openAPI) on the `/docs` endpoint

> * link:https://platform.openai.com/docs/api-reference/introduction[OpenAI API documentation]

Author: ```suggestion
* link:https://platform.openai.com/docs/api-reference/introduction[OpenAI API documentation^]
```

> * link:https://platform.openai.com/docs/api-reference/introduction[OpenAI API documentation]

Author: ```suggestion
* link:https://platform.openai.com/docs/api-reference/introduction[OpenAI API documentation]^
```
